### DISCUSSION TITLE
Update version counter when relationship updates

### DISCUSSION THREAD

**Discussion by petarlishov** [sqlalchemy/sqlalchemy#11887] [Usage Questions] [answered]
Hi,

I noticed that in a one to many relationship, the version counter on the parent table does not update when the child rows change. I was looking for a nice way to force the update of the version_id column only on the parent row then as well.

My use-case is that I want to consider a collection of data on a set of DB tables as belonging to one "business object" in Python where one Python object is the root and from it you can access the other children objects (e.g. if an order is a business object and you wanna have the root order object with the overall price but also a set of items inside it, where you want to have the order price based on the items inside it). Fetching, validations and operations are to apply collectively on all objects with the same root in Python. I want the versioning tick to happen on the root only once per transaction regardless of the object being modified (root or children). That way I can prevent concurrent changes to a collection of objects with the same root which could otherwise cause a bad state/validation misses on the whole "business object" (imagine an order that has the wrong overall price due to concurrent updates to the items inside it).

Example code:
```python
class Parent:
    id: int
    version_id: int
    children: list[Child]

    def __init__(children: list[Child]):
        if not children:
            children = []
        self.children = children

class Child:
    parent_id: int
    value: str

    def __init__(self, value: str):
        self.example = example


def _map(self, mapper: registry, schema: str):
    parent_table = Table(
        "parent",
        mapper.metadata,
        Column("id", Integer, nullable=False, primary_key=True),
        Column("version_id", Integer, nullable=False),
        schema=schema,
    )
    
    child_table = Table(
        "child",
        mapper.metadata,
        Column("parent_id", ForeignKey(parent_table.c.id), primary_key=True),
        Column("value", String(length=100), primary_key=True),
        schema=schema,
    )

    mapper.map_imperatively(Child, child_table)
    mapper.map_imperatively(Parent, parent_table, properties={"children": relationship(Child, cascade="all, delete-orphan")})
```

I am looking for a way to reliably increase the version counter only once per transaction and regardless of whether parts on the parent or child objects or their relationship change. 

I managed to find a way to detect some of the changes on the relationship with something like this:
```python
@event.listens_for(Parent.children, 'modified', propagate=True)
def on_children_modify(target, value, initiator):
    ...

@event.listens_for(Parent.children, 'remove', propagate=True)
def on_children_remove(target, value, initiator):
    ...

@event.listens_for(Parent.children, 'set', propagate=True)
def on_children_set(target, value, initiator):
    ...
```
I do not have an easy way to detect changes to the child row that do not pertain to the relationship though e.g. changing the child value and I also do not know how to force the version counter tick once I detect these events...

Basically, I am looking for a passing test that looks a bit like this:
```python
def test_persistence(engine):
    parent = Parent(children=[Child(value="1"), Child(value="2")])
    parent_id: int

    with Session(bind=engine) as session, session.begin():
        session.add(parent)
        session.flush()
        parent_id = parent.id

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        assert fetched.version_id == 1

        fetched.children.add(Child(value="3"))

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        # Ticks when a child addition happens
        assert fetched.version_id == 2

        fetched.children.remove(fetched.children[0])

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        # Ticks when a child removal happens
        assert fetched.version_id == 3

        fetched.children = [Child(value="4")]

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        # Ticks when a complete reassignment of children happens
        assert fetched.version_id == 4

        fetched.children[0].value = "5"

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        # Ticks when a change only on the child happens
        assert fetched.version_id == 5

        fetched.children.add(Child(value="6"))
        session.flush()
        fetched.children.remove(fetched.children[0])

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        # Ticks only by 1 when multiple changes happened
        assert fetched.version_id == 6

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(parent_id)
        # Does not tick when no change happened
        assert fetched.version_id == 6
```

I do not know how to trigger the version counter to reliably tick only once per transaction and in all the above cases. I would generally be keen on having a convenient way to set this on the relationship or the mapping of the child class somehow so that the behaviour is automatic and does not need setting of explicit event listeners, but can understand if you want to keep things as they are and not spin off any feature requests at the back of this.

**Comment by zzzeek** [comment, marked as answer]
I would scan for these objects in the [SessionEvents.before_flush](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.before_flush) event and increment the version counter in there.   if you want to make it once per transaction across flushes, you probably want to memoize the objects and the state of their version counter bump as well, which you could do on a separate variable on each object or using the `Session.info` dictionary.   You'd reset all this state when the transaction ends using the [SessionEvents.after_transaction_end](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_transaction_end) event.

**Reply by petarlishov** [reply]
Just to check, should it be that I use a manual version generator so that I bump it when I say and then I listen on before_commit to bump manually (not sure why I need to interact with after_transaction_end as it sounds like that happens after the deal is done)? I guess a commit is the same as a "before_transaction_end", but I may be very wrong. Also I guess in order to bump the version only by one, I would need to look at some sort of dirtiness boolean that is set during a transaction so that at the end before the commit only if the object is dirty do I bump the version? What does the session use to determine whether the object has changed and needs persisted? Can I use that to any extent or do I need to make something custom? I fear I might need to carefully monitor for flush events as well if the session does not track "dirtiness" across flushes?

**Reply by zzzeek** [reply]
just quick if you want to see an example of checking for dirtiness in before_flush you can see https://github.com/sqlalchemy/sqlalchemy/blob/main/examples/versioned_history/history_meta.py

the "commit" event is per-connection, which is usually the same as per-"transaction" unless you are have multiple engines in play.
